xop.go

```diff
@@ -79,86 +79,152 @@ func (d *xopDecoder) getXopContentIDIncludePath(element *etree.Element, path []s
+// Unwrap value as much as possible. A value can no longer be unwrapped if:
+// - it is an empty array or slice
+// - it is a nil pointer
+// This assumes, if it encounters an array field, that it will be filling in the first field and continuing
+// TODO: support arbitrary array size.
+func unwrapValue(val reflect.Value) reflect.Value {
+	// if the value is an interface or pointer, get its value
+	if val.Type().Kind() == reflect.Ptr || val.Type().Kind() == reflect.Interface {
+		// if the value is a nil pointer
+		if val.IsNil() {
+			return val
+		}
+
+		return unwrapValue(val.Elem())
 	}
 
-	// WARNING: Dragons below.
-	// We iterate the Content struct.
-	// We attempt to confirm that this is the proper object to store the data in, by
-	// examining the XMLName field and confirming that it matches the first element found in the XOP include path.
-	// Next, we attempt to find the field identified by the subsequent elements found in the XOP include path
-	// by looking for the XML tag that matches then repeating until we reach the end of the include path.
-	// We use HasPrefix, instead of ==, because often the tag includes ',omitempty' as a trailer.
-	// Once we've got it, we return the element.
-
-	rXMLName, err := getFieldByName(val.Elem(), "XMLName")
-	if err != nil {
-		return rXMLName, err
+	// if the value is an array or a slice, assume that we are looking for its first element
+	if val.Type().Kind() == reflect.Slice || val.Type().Kind() == reflect.Array {
+		// if the value is an empty array or slice
+		if val.Cap() == 0 {
+			return val
+		}
+
+		return unwrapValue(val.Index(0))
 	}
 
-	// The XMLName field has 2 properties; one called 'Space' and one called 'Local'.
-	// We only care about 'Local' right now.
-	// We assume the namespace is correct if the local name is correct.
-	for i := 0; i < rXMLName.NumField(); i++ {
-		rXMLTypeField := rXMLName.Type().Field(i)
-		rXMLValueField := rXMLName.Field(i)
+	// the value has been unwrapped
+	return val
+}
 
-		if rXMLTypeField.Name == "Local" && rXMLValueField.String() != path[1] {
-			return reflect.Value{}, fmt.Errorf("invalid XML object path, expected %s, got %s", rXMLValueField.String(), path[1])
-		}
+func getNameFromTag(tag string) string {
+	// if the tag is not set then there is no name to obtain
+	if tag == "" {
+		return ""
 	}
 
-	rValueField := val.Elem()
-	for i := 2; i < len(path); i++ {
-		rValueField, err = getFieldByXMLTagPrefix(rValueField, path[i])
-		if err != nil {
-			return rValueField, err
+	// remove xml namespace from the front of the tag
+	parts := strings.Split(tag, " ")
+	tag = parts[len(parts) - 1]
+
+	// return the XMLName from the front of the remaining tag
+	return strings.Split(tag, ",")[0]
+}
+
+// getExplicitXMLName gets the xml name which is explicitly set in the xml tag on the XMLName field
+func getExplicitXMLName(val reflect.Value) string {
+	// only a value of type reflect.Struct can have an XMLName field
+	if val.Type().Kind() == reflect.Struct {
```

Comment:
I personally prefer the early return approach, i.e. use:
```
if val.Type().Kind() != reflect.Struct {
  return ""
}
... other code ...
``` 
as it keeps the level of indentation down as you get into the body of the function.